### ISSUE TITLE
gimme-aws-creds not asking for Multi-factor Authentication configuration

### ISSUE THREAD

**Issue by nsharma-fy** [Nike-Inc/gimme-aws-creds#331]
## Expected Behavior
2 factors authentication is enabled in the AWS Federation OKTA app, running "gimme-aws-creds" should ask for MFA under "Multi-factor Authentication required." section. 

## Current Behavior

MFA information is not asked as seen below

```
$ gimme-aws-creds --version
gimme-aws-creds 2.4.3

$  gimme-aws-creds --register_device --profile DEFAULT
Okta Password for <okta-user>:
Device token saved!

$ gimme-aws-creds
Okta Password for <okta-user>:
Traceback (most recent call last):
  File "/home/ec2-user/.local/bin/gimme-aws-creds", line 17, in <module>
    GimmeAWSCreds().run()
  File "/home/ec2-user/.local/lib/python3.7/site-packages/gimme_aws_creds/main.py", line 469, in run
    self._run()
  File "/home/ec2-user/.local/lib/python3.7/site-packages/gimme_aws_creds/main.py", line 802, in _run
    for data in self.iter_selected_aws_credentials():
  File "/home/ec2-user/.local/lib/python3.7/site-packages/gimme_aws_creds/main.py", line 773, in iter_selected_aws_credentials
    for role in self.aws_selected_roles:
  File "/home/ec2-user/.local/lib/python3.7/site-packages/gimme_aws_creds/main.py", line 672, in aws_selected_roles
    selected_roles = self._get_selected_roles(self.requested_roles, self.aws_roles)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/gimme_aws_creds/main.py", line 663, in aws_roles
    self.saml_data['SAMLResponse'],
  File "/home/ec2-user/.local/lib/python3.7/site-packages/gimme_aws_creds/main.py", line 654, in saml_data
    self._cache['saml_data'] = saml_data = self.okta.get_saml_response(self.aws_app['links']['appLink'])
  File "/home/ec2-user/.local/lib/python3.7/site-packages/gimme_aws_creds/okta.py", line 725, in get_saml_response
    raise RuntimeError(saml_error)
RuntimeError: Did not receive SAML Response after successful authentication [<app_url>]

$ cat ~/.okta_aws_login_config
[DEFAULT]
okta_org_url = <org_url>
okta_auth_server =
client_id =
gimme_creds_server = appurl
aws_appname =
aws_rolename = all
write_aws_creds = True
cred_profile = role
okta_username = <okta_user>
app_url = <app_url>
resolve_aws_alias = False
include_path = False
preferred_mfa_type = token:software:totp
remember_device = False
aws_default_duration = 3600
output_format =
device_token = <token>
```

The utility works fine with single-factor authentication. I suspect some configuration issue.  MFA (using Okta verify) works for the user at Okta console.

**Comment by roboweaver** [comment]
I get this message too when trying to use the new OIE Okta Identity Engine flow - been talking to Okta and it appears there is some change in the SAML flow with this rule enabled:

![image](https://user-images.githubusercontent.com/3752123/145892320-318d8555-8d48-40f2-96e6-9e5bd1c68ad6.png)

**Comment by nsharma-fy** [comment]
I see slightly different options in sign on rules in my AWS Account Federation app
<img width="916" alt="Screen Shot 2021-12-13 at 4 05 13 PM" src="https://user-images.githubusercontent.com/79729374/145908794-9db9aa4f-6fc3-40aa-97a4-76094c50e050.png">

**Comment by roboweaver** [comment]
On the old flow in the browser, Okta prompts for login and a form is displayed for the auth type. On the new flow a screen is displayed that shows the options for login like this, and the code appears to be looking for a form field of "SAMLResponse" in the HTML:

<img width="543" alt="image002" src="https://user-images.githubusercontent.com/3752123/146024701-8cb027fb-9f40-4c79-8ca6-ec4a82711335.png">

**Comment by roboweaver** [comment]
Just for ducks, I changed my `preferred_mfa_type` to be empty (which in the old flow would prompt me to pick one), and very similar error, so it is seems to me this code is trying to parse the page with the MFA options and handle that form instead of using an API to do that part.

`Using password from keyring for robw@adobe.com
Traceback (most recent call last):
  File "/usr/local/bin/gimme-aws-creds", line 17, in <module>
    GimmeAWSCreds().run()
  File "/usr/local/lib/python3.9/site-packages/gimme_aws_creds/main.py", line 469, in run
    self._run()
  File "/usr/local/lib/python3.9/site-packages/gimme_aws_creds/main.py", line 802, in _run
    for data in self.iter_selected_aws_credentials():
  File "/usr/local/lib/python3.9/site-packages/gimme_aws_creds/main.py", line 773, in iter_selected_aws_credentials
    for role in self.aws_selected_roles:
  File "/usr/local/lib/python3.9/site-packages/gimme_aws_creds/main.py", line 672, in aws_selected_roles
    selected_roles = self._get_selected_roles(self.requested_roles, self.aws_roles)
  File "/usr/local/lib/python3.9/site-packages/gimme_aws_creds/main.py", line 663, in aws_roles
    self.saml_data['SAMLResponse'],
  File "/usr/local/lib/python3.9/site-packages/gimme_aws_creds/main.py", line 654, in saml_data
    self._cache['saml_data'] = saml_data = self.okta.get_saml_response(self.aws_app['links']['appLink'])
  File "/usr/local/lib/python3.9/site-packages/gimme_aws_creds/okta.py", line 725, in get_saml_response
    raise RuntimeError(saml_error)
RuntimeError: Did not receive SAML Response after successful authentication [https://adobewf-aws-dev.okta.com/home/amazon_aws/0oa8q5b01gPJrTUnJ696/272]`

**Comment by carlalexander** [comment]
Been struggling with this and talked with an Okta engineer. What you need to get the CLI to work is to have `Require secondary factor` checked. This will fix the issue.

<img width="600" alt="image" src="https://user-images.githubusercontent.com/654684/166075830-4a350ab6-207a-4b9d-a8c9-b71428afdda7.png">

Without it, Okta tells that the initial authentication by gimme-aws-creds was successful when in fact it wasn't. So the cli goes ahead tries to fetch the `appurl` and fails because it's not actually logged in properly. With that option checked, the initial authencation will return `MFA_REQUIRED` and the authentication flow will continue normally.

**Comment by nsharma-fy** [comment]
@carlalexander, my rule editor UI is different from yours, https://github.com/Nike-Inc/gimme-aws-creds/issues/331#issuecomment-993024655 and doesn't have the "Require secondary factor" checkbox 

What kind of application you have, I have "AWS Account Federation"

**Comment by ataylorme** [comment]
@nsharma-fy the configuration is not in the AWS Federation App. You must create a new global session policy in the Okta admin to require MFA, then assign that global session policy to user groups who have the AWS Federation App assigned to them


<img width="1339" alt="okta-oie-global-session-policy-screenshot" src="https://user-images.githubusercontent.com/2133004/166517303-5e33d7d3-a66c-4225-a292-a01d53774139.png">
.

**Comment by nsharma-fy** [comment]
Thanks @ataylorme , I will give it a try.

**Comment by seanorama** [comment]
We experienced the same:

1. The Default "Authentication Policy" required MFA.
2. But the "Global Session Policy" did not.
3. Clients authenticated successfully with "password" but were never told they needed to send MFA.
4. After requiring MFA in the "Global Session Policy" we are able to authenticate.

This impacted other clients, such as `saml2aws`.

This is a bug on Okta side, since the `Authentication Policy` is a layer on top of the `Global Session Policy`. If MFA is required in the `Authentication Policy` then Okta should prompt for it.

**Comment by epierce** [comment]
OIE support using Okta's recommended flow (device authorization) was released in v2.6.0
